File: packages/kotlin/src/test/kotlin/com/luxo/client/CodegenTest.kt
package com.luxo.client

import kotlin.test.Test
import kotlin.test.assertContains
import kotlin.test.assertEquals
import kotlin.test.assertFalse

class CodegenTest {
    private val schema = LuxoSchema(
        models = mapOf(
            "Payload" to LuxoModel(
                name = "Payload",
                usage = TypeUsage.OUTPUT,
                fields = listOf(
                    LuxoField(1, "id", "Int"),
                    LuxoField(2, "blob", "Bytes"),
                    LuxoField(3, "metadata", "JSON"),
                ),
            ),
        ),
        types = mapOf(
            "CreateInput" to LuxoTypeDecl(
                name = "CreateInput",
                usage = TypeUsage.INPUT,
                fields = listOf(LuxoField(1, "name", "String")),
            ),
        ),
        apis = mapOf(
            "createPayload" to LuxoAPI(
                id = 6,
                name = "createPayload",
                module = "file",
                returnType = "Payload",
                params = listOf(
                    LuxoParam(1, "input", "JSON", typeName = "CreateInput"),
                    LuxoParam(2, "note", "String", nullable = true),
                    LuxoParam(3, "caption", "String", nullable = true, hasDefault = true),
                ),
            ),
            "listPayloads" to LuxoAPI(
                id = 7,
                name = "listPayloads",
                module = "file",
                returnType = "Payload",
                returnList = true,
            ),
            "listPayloadPage" to LuxoAPI(
                id = 8,
                name = "listPayloadPage",
                module = "file",
                returnType = "Payload",
                returnList = true,
                paginated = true,
            ),
            "watchPayload" to LuxoAPI(
                id = 9,
                name = "watchPayload",
                module = "file",
                returnType = "Payload",
                stream = true,
                params = listOf(LuxoParam(1, "projectId", "Int")),
            ),
            "searchPayloads" to LuxoAPI(
                id = 10,
                name = "searchPayloads",
                module = "file",
                returnType = "Payload",
                returnList = true,
                paginated = true,
            ),
            "createSnapshot" to LuxoAPI(
                id = 11,
                name = "createSnapshot",
                module = "file",
                returnType = "Payload",
            ),
        ),
    )

    @Test
    fun `introspection key is sent only in the canonical header`() {
        val connection = LuxoCodegen.introspectionConnection("https://api.example.com/luvia", "secret-key")

        assertEquals("secret-key", connection.getRequestProperty("X-Introspection-Key"))
        assertEquals("\$schema", connection.url.query)
        assertFalse(connection.url.toString().contains("key="))
        connection.disconnect()
    }

    @Test
    fun `generates canonical columnar and paginated decoders`() {
        val types = LuxoCodegen.genTypes(schema, "com.example")
        val client = LuxoCodegen.genClient(schema, "com.example")
        val hints = LuxoCodegen.genSelectHints("com.example", emptyMap())

        assertContains(types, "fun decodeColumnarPayload(data: ByteArray): List<Payload>")
        assertContains(types, "fun decodePaginatedPayload(data: ByteArray): Page<Payload>")
        assertContains(types, "import com.luxo.client.ColumnarDecoder")
        assertContains(types, "import com.luxo.client.Page")
        assertContains(types, "val id: Selected<Long> = Selected.unselected()")
        assertContains(types, "Selected.value(dec.readInt())")
        assertFalse(types.contains("?: 0L"))
        assertContains(types, "fun decodeJSONPayload(data: JsonObject): Payload")
        assertContains(types, "Base64.getDecoder().decode")
        assertContains(types, "dec.readColumnBytes()")
        assertContains(types, "Json.parseToJsonElement")
        assertFalse(types.contains("!!"))
        assertContains(client, "decodeColumnarPayload(data)")
        assertContains(client, "decodePaginatedPayload(data)")
        assertContains(client, "input: CreateInput")
        assertContains(client, "Json.encodeToJsonElement(input)")
        assertContains(client, "note: String?")
        assertContains(client, "caption: LuxoOptional<String> = LuxoOptional.Absent")
        assertContains(client, "is LuxoOptional.Present -> callParams[\"caption\"] = caption.value")
        assertContains(client, "filters: List<LuxoFilter>? = null")
        assertContains(client, "callParams[\"\\\$filters\"] = filters.map")
        assertContains(client, "suspend fun subscribeWatchPayload(projectId: Long, select: String? = null, onData: (Payload) -> Unit): () -> Unit")
        assertContains(client, "transport.subscribe(\"watchPayload\", callParams)")
        assertContains(client, "onData(decode_watchPayload(data))")
        assertContains(client, "endpoint.startsWith(\"ws://\") || endpoint.startsWith(\"wss://\")")
        assertContains(client, "LuxoWebSocket(endpoint, token = token)")
        assertContains(client, "suspend fun searchPayloads(page: Long? = null, pageSize: Long? = null, select: String? = null, filters: List<LuxoFilter>? = null, sorters: List<LuxoSorter>? = null)")
        assertContains(client, "suspend fun createSnapshot(select: String? = null): Payload")
        assertFalse(client.contains("createSnapshot(input:"))
        assertContains(client, "suspend fun createPayload(input: CreateInput, note: String?, caption: LuxoOptional<String> = LuxoOptional.Absent, select: String? = null)")
        assertContains(hints, "object SelectHints")
        assertContains(hints, "emptyMap()")
        assertFalse(client.contains("try { SelectHints.hints[api] }"))
        assertFalse(client.contains("!!"))
    }

    @Test
    fun `separates strict input DTOs from selected output models`() {
        val shared = LuxoSchema(
            models = emptyMap(),
            types = mapOf(
                "Profile" to LuxoTypeDecl(
                    name = "Profile",
                    usage = TypeUsage.UNUSED,
                    fields = listOf(
                        LuxoField(1, "name", "String"),
                        LuxoField(2, "bio", "String", nullable = true),
                    ),
                ),
            ),
            apis = mapOf(
                "updateProfile" to LuxoAPI(
                    id = 1,
                    name = "updateProfile",
                    module = "profile",
                    returnType = "Profile",
                    params = listOf(LuxoParam(1, "profile", "Model", typeName = "Profile")),
                ),
            ),
        )

        val types = LuxoCodegen.genTypes(shared, "com.example")
        val client = LuxoCodegen.genClient(shared, "com.example")

        assertContains(types, "data class Profile(")
        assertContains(types, "val name: Selected<String> = Selected.unselected()")
        assertContains(types, "val bio: Selected<String?> = Selected.unselected()")
        assertContains(types, "@Serializable\ndata class ProfileInput(")
        assertContains(types, "val name: String")
        assertContains(types, "val bio: String? = null")
        assertContains(client, "profile: ProfileInput")
        assertContains(client, "): Profile")
    }
}

File: packages/dart/test/codegen_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:luxo_client/src/codegen.dart';
import 'package:test/test.dart';

void main() {
  test('generated client preserves binary list, Bytes, JSON, and field IDs',
      () async {
    final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
    final output = Directory('.tmp/codegen_test');
    if (output.existsSync()) output.deleteSync(recursive: true);

    final requestFuture = server.first.then((request) async {
      expect(request.uri.queryParameters.containsKey(r'$schema'), isTrue);
      expect(request.uri.queryParameters.containsKey('key'), isFalse);
      expect(request.headers.value('X-Introspection-Key'), 'test');
      request.response.headers.contentType = ContentType.json;
      request.response.write(jsonEncode(_schema));
      await request.response.close();
    });

    try {
      await generate(
        endpoint: 'http://${server.address.host}:${server.port}/luxo',
        key: 'test',
        outDir: output.path,
      );
      await requestFuture;

      final types = File('${output.path}/types.dart').readAsStringSync();
      final schema = File('${output.path}/schema.dart').readAsStringSync();
      final client = File('${output.path}/client.dart').readAsStringSync();

      expect(types, contains("import 'dart:typed_data';"));
      expect(types, contains('final Selected<Uint8List> data;'));
      expect(types, contains('final Selected<Object> metadata;'));
      expect(types, contains('final Selected<List<Uint8List>> chunks;'));
      expect(types, contains('final Selected<Child> child;'));
      expect(types, contains('class ChildInput'));
      expect(types, contains('final ChildInput child;'));
      expect(types, contains('Child.fromJson'));
      expect(types, contains("if (data.isSelected) 'data': base64Encode(data.value)"));
      expect(types, contains("if (child.isSelected) 'child': child.value.toJson()"));
      expect(types, isNot(contains('Uint8List(0)')));
      expect(types, contains('dec.readBytes()'));
      expect(types, contains('jsonDecode(utf8.decode(dec.readBytes()))'));
      expect(
        schema,
        contains("ParamSchema(1, 'chunks', 'Bytes', true, false)"),
      );
      expect(
        schema,
        contains("'child': SelectionFieldSchema(4, 'Child')"),
      );
      expect(schema,
          contains("luxoSelectionTypes['Payload']!, luxoSelectionTypes"));
      expect(client, contains('required List<Uint8List> chunks'));
      expect(client, contains('required Object metadata'));
      expect(client, contains('required CreateInput input'));
      expect(client, contains('required String? note'));
      expect(
        client,
        contains(
          'LuxoOptional<String> caption = const LuxoOptional<String>.absent()',
        ),
      );
      expect(
          client, contains("if (caption.isPresent) 'caption': caption.value"));
      expect(client, isNot(contains('..nextField()')));
      expect(types,
          contains('List<Payload> decodeColumnarPayload(Uint8List data)'));
      expect(types,
          contains('Page<Payload> decodePaginatedPayload(Uint8List data)'));
      expect(client, contains('decodeColumnarPayload(d)'));
      expect(client, contains('decodePaginatedPayload(d)'));
      expect(client, contains('List<LuxoFilter>? filters'));
      expect(client, contains("if (filters != null) r'\$filters': filters"));
      expect(
        client,
        contains(
          'Future<Page<Payload>> searchPayloads({int? page, int? pageSize, String? select, List<LuxoFilter>? filters, List<LuxoSorter>? sorters})',
        ),
      );
      expect(
        client,
        contains('Future<Payload> createSnapshot({String? select})'),
      );
      expect(client,
          isNot(contains('createSnapshot(Map<String, dynamic> input)')));
      expect(client, contains('Future<Payload> upload({'));
      expect(client, contains('String? select'));
      expect(
        client,
        contains(
          'Future<void Function()> subscribeWatchPayload({required int projectId, String? select, required void Function(Payload) onData})',
        ),
      );
      expect(client, contains("transport.subscribe('watchPayload'"));
      expect(
        client,
        contains(
          'onData(d is Uint8List ? decodePayload(LuxoDecoder(d)) : Payload.fromJson(d as Map<String, dynamic>))',
        ),
      );

      final analysis = await Process.run(
        Platform.resolvedExecutable,
        ['analyze', output.path],
      );
      expect(
        analysis.exitCode,
        0,
        reason: '${analysis.stdout}\n${analysis.stderr}',
      );
    } finally {
      await server.close(force: true);
      if (output.existsSync()) output.deleteSync(recursive: true);
    }
  });
}

const _schema = <String, Object>{
  'models': {
    'Child': {
      'name': 'Child',
      'usage': 'unused',
      'fields': [
        {'id': 1, 'name': 'name', 'type': 'String'},
      ],
    },
    'Payload': {
      'name': 'Payload',
      'usage': 'output',
      'fields': [
        {'id': 1, 'name': 'data', 'type': 'Bytes'},
        {'id': 2, 'name': 'metadata', 'type': 'JSON'},
        {'id': 3, 'name': 'chunks', 'type': 'Bytes', 'isList': true},
        {
          'id': 4,
          'name': 'child',
          'type': 'Model',
          'typeName': 'Child',
          'relation': true,
        },
      ],
    },
    'CreateInput': {
      'name': 'CreateInput',
      'usage': 'input',
      'fields': [
        {'id': 1, 'name': 'child', 'type': 'Model', 'typeName': 'Child'},
      ],
    },
  },
  'apis': {
    'upload': {
      'id': 7,
      'name': 'upload',
      'module': 'file',
      'returnType': 'Payload',
      'params': [
        {'id': 1, 'name': 'chunks', 'type': 'Bytes', 'isList': true},
        {'id': 2, 'name': 'metadata', 'type': 'JSON'},
        {'id': 3, 'name': 'input', 'type': 'JSON', 'typeName': 'CreateInput'},
        {'id': 4, 'name': 'note', 'type': 'String', 'nullable': true},
        {
          'id': 5,
          'name': 'caption',
          'type': 'String',
          'nullable': true,
          'hasDefault': true,
        },
      ],
    },
    'listPayloads': {
      'id': 8,
      'name': 'listPayloads',
      'module': 'file',
      'returnType': 'Payload',
      'returnList': true,
    },
    'listPayloadPage': {
      'id': 9,
      'name': 'listPayloadPage',
      'module': 'file',
      'returnType': 'Payload',
      'returnList': true,
      'paginated': true,
    },
    'watchPayload': {
      'id': 10,
      'name': 'watchPayload',
      'module': 'file',
      'returnType': 'Payload',
      'stream': true,
      'params': [
        {'id': 1, 'name': 'projectId', 'type': 'Int'},
      ],
    },
    'searchPayloads': {
      'id': 11,
      'name': 'searchPayloads',
      'module': 'file',
      'returnType': 'Payload',
      'returnList': true,
      'paginated': true,
    },
    'createSnapshot': {
      'id': 12,
      'name': 'createSnapshot',
      'module': 'file',
      'returnType': 'Payload',
    },
  },
};
